### ISSUE TITLE
Configuration file in TOML?

### ISSUE THREAD

**Issue by ueno** [keylime/keylime#895]
In the current keylime configuration, I see a few inconsistencies in representing constructed types such as lists:
```
severity_labels = ["info", "notice", "warning", "error", "critical", "alert", "emergency"]
...
accept_tpm_hash_algs = sha512,sha384,sha256,sha1
```
The former is parsed using the python `ast` module, while the latter is manually split (whitespaces around comma are sometimes tolerated, but sometimes not). Perhaps it might make sense to eventually migrate to a more expressive language like [TOML](https://toml.io/en/) for usability and portability across languages (i.e., python and rust)?

**Comment by THS-on** [comment]
I really like the idea. There are a  few things to consider:

- The Python code currently relies on ConfigParser and that the logging config is in that INI format. This will require some work to change that.
- The stacking for config files should still work.
- Upgrade path for existing configs. Can we provide a way for either converting or still loading old Python config files?
- Adding an `include` keyword to start adressing https://github.com/keylime/keylime/issues/441
- Start versioning our config file(s), so that we can potentially implement automatic migrations to newer version with sane defaults.

**Comment by mpeters** [comment]
All of these ideas (including using TOML) all sound good to me, but are at least semi-disruptive. I'm definitely open to having an enhancement document we can kick around and design a future solution with.

**Comment by THS-on** [comment]
@ueno can you open a enhancement proposal for this change?

**Comment by ueno** [comment]
> @ueno can you open a enhancement proposal for this change?

Sorry I haven't had time to do that in the past couple of weeks, but probably @ansasaki will help move this forward with his experience in certificate configuration :-) A few minor comments from the implementation perspective:
- It would be nice if a lint program is provided for the new format; maybe based on [JSON schema on TOML](https://json-schema-everywhere.github.io/toml)
- As for versioning, maybe the file extension could be used to distinguish the initial version and the later versions; the new format could have a version attribute natively in the file

**Comment by mpeters** [comment]
> * As for versioning, maybe the file extension (`*.conf` vs `*.toml`) could be used to distinguish the initial version and the later versions; the new format could have a version attribute natively in the file

I'd prefer to keep the file extension the same. Maybe we try to parse as TOML first and then if that passes we assume it's TOML? Is the current config accidentally valid TOML? Or some other way to determine?

I agree about the version attribute in the file though.

**Comment by THS-on** [comment]
We have chosen not switch to TOML for the Python components. The rest is implemented in #1068.
